pages/en/lb4/Testing-Your-application.md

```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
+
+A typical unit-test creates a controller instance with dependencies replaced by Test doubles and directly calls the tested method. In the example below, we give the controller a stub implementation of its repository dependency, see [Create a stub repository](#create-a-stub-repository) for a detailed explanation.
+
+```ts
+// test/controllers/product.controller.unit.ts
+import {ProductController, ProductRepository} from '../..';
+import {expect, sinon} from '@loopback/testlab';
+
+describe('ProductController (unit)', () => {
+  let repository: ProductRepository;
+  beforeEach(givenStubbedRepository);
+
+  describe('getDetails()', () => {
+    it('retrieves details of a product', async () => {
+      const controller = new ProductController(repository);
+      const findStub = repository.find as sinon.SinonStub;
+      findStub.resolves([{id: 1, name: 'Pen'}]);
+
+      const details = await controller.getDetails(1);
+
+      expect(details).to.containDeep({name: 'Pen'});
+      expect(findStub).to.be.calledWithMatch({where: {id: 1}});
+    });
+  });
+
+  function givenStubbedRepository() {
+    repository = sinon.createStubInstance(ProductRepository);
+  }
+});
+```
+
+### Unit-test your Models and Repositories
+
+In a typical LoopBack appliction, models and repositories are relying mostly on the behavior provided by the framework (`@loopback/repository` package) and there is no need to test the LoopBack's built-in functionality unit-test level (the functionality has been already verified by the tests in the framework). What does need new unit-tests is any additional application-specific API.
+
+For example, if our `Person` Model has properties `firstname`, `middlename` and `surname` and provides a function to obtain the full name, then we should write unit tests to verify the implementation of this additional method.
+
+```ts
+// test/unit/models/person.model.unit.ts
+import {Person} from '../../models/person.model.ts'
+import {expect} from '@loopback/testlab';
+
+describe('Person (unit)', () => {
+  // we recommend to group tests by method names
+  describe('getFullName()', () => {
+    it('uses all three parts when present', () => {
+      const person = new Person({
+        firstname: 'Jane',
+        middlename: 'Smith',
+        surname: 'Brown'
+      });
+
+      const fullName = person.getFullName();
+
+      expect(fullName).to.equal('Jane Smith Brown');
+    });
+
+    it('omits middlename when not present', () => {
+      // etc.
+    });
+  });
+});
+```
+
+Writing a unit test for a custom Repository methods is not straightforward because `CrudRepositoryImpl` is based on legacy loopback-datasource-juggler that was not designed with Dependency Injection in mind. We are recommending to use integration tests to verify the implementation of custom Repository methods, see [Test your Repositories against a real database](#test-your-repositories-against-a-real-database) in [Integration Testing](#integration-testing).
```

Comment:
Break up into separate lines.

Reply:
`Writing a unit test for a custom Repository methods`
should either lose the plurality on "methods" or remove the leading "a" from "a custom".